apis/python/src/tiledbsoma/collection.py

```diff
@@ -168,12 +176,70 @@ def add_new_collection(
         self._close_stack.enter_context(new_group)
         return new_group
 
-    def _not_implemented(self, *args: Any, **kwargs: Any) -> NoReturn:
-        raise NotImplementedError()
+    def add_new_dataframe(
+        self,
+        key: str,
+        *,
+        uri: Optional[str] = None,
+        schema: pa.Schema,
+        index_column_names: Sequence[str] = (SOMA_JOINID,),
+        platform_config: Optional[options.PlatformConfig] = None,
+    ) -> DataFrame:
+        if key in self:
+            raise KeyError(f"{key!r} already exists in {type(self)}")
+        self._check_allows_child(key, DataFrame)
+        absolute_uri, was_relative = self._new_absolute_uri(key=key, uri=uri)
+        new_df = DataFrame.create(
+            absolute_uri,
+            index_column_names=index_column_names,
+            schema=schema,
+            platform_config=platform_config,
+            context=self.context,
+        )
+        # A dataframe might not be the declared type of this collection,
+        # but we can't really handle that within the type system.
+        self._set_element(
+            key, new_df, use_relative_uri=was_relative  # type: ignore[arg-type]
+        )
+        self._close_stack.enter_context(new_df)
+        return new_df
 
-    add_new_dataframe = _not_implemented
-    add_new_dense_ndarray = _not_implemented
-    add_new_sparse_ndarray = _not_implemented
+    def _add_new_ndarray(
+        self,
+        cls: Type[_NDArr],
+        key: str,
+        *,
+        uri: Optional[str] = None,
+        type: pa.DataType,
+        shape: Sequence[int],
+        platform_config: Optional[options.PlatformConfig] = None,
+    ) -> _NDArr:
+        if key in self:
+            raise KeyError(f"{key!r} already exists in {type(self)}")
+        self._check_allows_child(key, cls)
+        absolute_uri, was_relative = self._new_absolute_uri(key=key, uri=uri)
+        new_arr = cls.create(
+            absolute_uri,
+            type=type,
+            shape=shape,
+            platform_config=platform_config,
+            context=self.context,
+        )
+        # An NDArray might not be the declared type of this collection,
+        # but we can't really handle that within the type system.
+        self._set_element(
+            key, new_arr, use_relative_uri=was_relative  # type: ignore[arg-type]
+        )
+        self._close_stack.enter_context(new_arr)
+        return new_arr
+
+    # These are both correct but mypy doesn't get it.
+    add_new_dense_ndarray = functools.partialmethod(  # type: ignore[assignment]
```

Comment:
These look exposed -- they're called `add_new_dense_ndarray` and `add_new_sparse_nd_array`, not `_add_new_dense_ndarray` and `_add_new_sparse_nd_array`. But they lack docstrings. They're only used in this PR as called from unit-test files which suggests to me they're user-facing.

There's Python magic going on here which as a non-Python-expert person I don't immediately grasp. But the reader shouldn't have to be a Python expert to grasp what we're doing here.

Please dedicate more wording to explaining, for the rest of us, what it _is_ that's _correct_ and what _is_ the "it" that mypy doesn't get it. Because I don't get it either.

Reply:
Added some commentary.

Reply:
@thetorpedodog thanks for addressing one of my questions!

Why are there not docstrings? Please add them or make clear why these apparently-user-facing methods don't actually need docstrings.

Reply:
The docstrings, while they are not literally inherited or available from the `__doc__` member, can be extracted by tooling from the root `somacore.BaseCollection` object. However, putting them there will be a cue that these are indeed the user-facing functions you're looking for.

Reply:
@thetorpedodog please create a tracking issue to make sure we don't forget, and link to it here when you've done so.